File: Leetcode/PartitionEqualSubsetSum.c
// https://leetcode.com/problems/partition-equal-subset-sum

#define MAX_DIF 100 * 200

int dp_[2 * MAX_DIF + 1] = {0};
int ndp_[2 * MAX_DIF + 1] = {0};

int canPartition(int* nums, int numsSize) {
    int MAX = 0;
    for (int i = 0; i < numsSize; ++i) {
        MAX += nums[i];
    }

    for (int i = 0; i < 2 * MAX + 1; ++i) {
        dp_[i] = ndp_[i] = 0;
    }

    int* dp = dp_;
    int* ndp = ndp_;

    dp[MAX] = 1;
    for (int i = 0; i < numsSize; ++i) {
        for (int i = 0; i < 2 * MAX + 1; ++i) {
            ndp[i] = 0;
        }

        for (int dif = 0; dif <= 2 * MAX; ++dif) {
            if (dif - nums[i] >= 0) {
                ndp[dif - nums[i]] |= dp[dif];
            }

            if (dif + nums[i] <= 2 * MAX) {
                ndp[dif + nums[i]] |= dp[dif];
            }
        }

        {
            int* t = dp;
            dp = ndp;
            ndp = t;
        }
    }

    return dp[MAX];
}

File: Leetcode/PartitionEqualSubsetSum.rs
// https://leetcode.com/problems/partition-equal-subset-sum

struct Solution {}

impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        const MAX_DIF: usize = 100 * 200;

        let mut dp = [false; 2 * MAX_DIF + 1];
        let mut ndp = [false; 2 * MAX_DIF + 1];

        dp[MAX_DIF] = true;
        for num in nums {
            let num = num as usize;

            for i in 0..=2 * MAX_DIF {
                ndp[i] = false;
            }

            for dif in 0..=2 * MAX_DIF {
                if dif >= num {
                    ndp[dif - num] |= dp[dif];
                }
                if dif + num <= 2 * MAX_DIF {
                    ndp[dif + num] |= dp[dif];
                }
            }

            std::mem::swap(&mut dp, &mut ndp);
        }

        dp[MAX_DIF]
    }
}
